File: Piece of cake/103. Binary Tree Zigzag Level Order Traversal/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
   public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root) return {};
        queue<TreeNode*> que;
        vector<vector<int>> res;
        que.push(root);
        while (!que.empty()) {
            size_t n{que.size()};
            res.emplace_back();
            while (n--) {
                TreeNode* cur{que.front()}; que.pop();
                res.back().emplace_back(cur->val);
                for (TreeNode* nxt : {cur->left, cur->right})
                    if (nxt != nullptr) que.push(nxt);
            }
        }
        for (size_t i{0}; i < res.size(); ++i)
            reverse(res[i].begin(), res[i].end());
        return res;
    }
};

int main() { return 0; }

File: Piece of cake/103. Binary Tree Zigzag Level Order Traversal/main.rs
#![allow(dead_code)]
struct Solution;
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;
impl Solution {
    pub fn zigzag_level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        if root.is_none() {
            return vec![];
        }
        let mut res = Vec::new();
        let mut que = VecDeque::new();
        que.push_back(root.unwrap());
        while !que.is_empty() {
            let mut v = Vec::new();
            for _ in 0..que.len() {
                let cur = que.pop_front().unwrap();
                v.push(cur.borrow().val);
                if let Some(left) = cur.borrow_mut().left.take() {
                    que.push_back(left);
                };
                if let Some(right) = cur.borrow_mut().right.take() {
                    que.push_back(right);
                };
            }
            res.push(v);
        }
        res.iter_mut()
            .skip(1)
            .step_by(2)
            .for_each(|row| row.reverse());
        res
    }
}

fn main() {}
